cairo/ethereum/cancun/vm/instructions/system.cairo

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+
+from ethereum.cancun.vm.stack import pop
+from ethereum.cancun.vm import Evm, EvmImpl
+from ethereum.cancun.vm.exceptions import Revert, OutOfGasError
+from ethereum.cancun.vm.memory import memory_read_bytes, expand_by
+from ethereum.cancun.vm.gas import calculate_gas_extend_memory, charge_gas, GasConstants
+from ethereum_types.numeric import U256, Uint, bool
+from ethereum.exceptions import EthereumException
+
+from ethereum_types.others import (
+    ListTupleU256U256,
+    ListTupleU256U256Struct,
+    TupleU256U256,
+    TupleU256U256Struct,
+)
+// @notice Revert operation - stop execution and revert state changes, returning data from memory
+func revert{range_check_ptr, evm: Evm}() -> EthereumException* {
+    alloc_locals;
+    // STACK
+    let stack = evm.value.stack;
+    with stack {
+        let (memory_start_index, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+        let (size, err) = pop();
+        if (cast(err, felt) != 0) {
+            return err;
+        }
+    }
+
+    // GAS
+    // Calculate memory expansion cost
+    // If size > 2**128 - 1, OutOfGasError
+    if (size.value.high != 0) {
+        tempvar err = new EthereumException(OutOfGasError);
+        return err;
+    }
+    tempvar extensions_tuple = new TupleU256U256(new TupleU256U256Struct(memory_start_index, size));
+    tempvar extensions_list = ListTupleU256U256(new ListTupleU256U256Struct(extensions_tuple, 1));
+    let extend_memory = calculate_gas_extend_memory(evm.value.memory, extensions_list);
+
+    let err = charge_gas(Uint(extend_memory.value.cost.value));
+    if (cast(err, felt) != 0) {
+        return err;
+    }
+
+    // OPERATION
+    let memory = evm.value.memory;
+    with memory {
+        // Extend memory
+        expand_by(extend_memory.value.expand_by);
+        // Read output from memory
+        let output = memory_read_bytes(memory_start_index, size);
+        // Set output
+        EvmImpl.set_output(output);
+    }
+
+    // Raise revert
+    tempvar revert = new EthereumException(Revert);
+    return revert;
+}
```

Comment:
missing stack re-bind?

Reply:
wait actually im realising we might need to re-bind the stacks right before returning _on every occurence_ and not at the end of functions

imagine: you pop `size` from the stack, your charge_gas tells your you're OOG, but the prover just sent you a wrong size, and because you didnt re-bind stack before returning, you're not squashing the last dict access, which makes your tx unsound

Reply:
yep sounds accurate

Reply:
very good point